### ISSUE TITLE
[Bug]: DeepSeek-V4-Flash-0731 DSpark K5 emits intermittent cross-script token corruption during long C6 generation

### ISSUE THREAD

**Issue by devinkuhn** [local-inference-lab/vllm#478]
## Summary

DeepSeek-V4-Flash-0731 on 4× RTX PRO 6000 Blackwell (SM120), TP4/DCP1, B12X sparse MLA, FP8 KV cache, LMCache, prefix caching, and DSpark K5 intermittently emits new cross-script token corruption during concurrent long-context generation.

The failure is silent at the serving layer: affected requests finish normally and the engine remains running. It is not a client SSE-reassembly problem; the foreign/glued tokens are present in captured raw SSE and parsed deltas.

We reproduced the corruption twice with the same retained input and runtime after target-only tests had failed to reproduce it under substantially heavier cache churn.

This issue is related to, but distinct from:

- local-inference-lab/rtx6kpro#53 — batch-wide BOS repetition with acceptance collapse to 0%
- local-inference-lab/rtx6kpro#72 — K7 multilingual word salad at C2 with acceptance collapse to 5.3%
- vllm-project/vllm#50720 — DSpark SM120 sparse-MLA warmup dispatch crash
- vllm-project/vllm#52938 — target-only SM120 sparse-MLA frozen logits / identical BOS
- local-inference-lab/vllm#475 — r19 isolated long-reasoning repetition at C4

The distinguishing result here is **repeatable new multilingual/glued-token corruption on r18 DSpark K5 at C6, while aggregate DSpark acceptance remains approximately 45% rather than collapsing globally**.

### Host

```text
OS: Ubuntu 24.04.4 LTS
Kernel: 6.8.0-136-generic
CPU: AMD Ryzen Threadripper 9970X, 32 cores / 64 threads
RAM: single NUMA node
GPU: 4× NVIDIA RTX PRO 6000 Blackwell Max-Q Workstation Edition, 97,887 MiB each
Topology: all GPUs NODE-connected, no NVLink
Driver: 595.71.05
CUDA runtime: 13.3.73
```

### Runtime image and source identity

```text
Image digest:
voipmonitor/vllm@sha256:414ec7d0d28358cfd8af0697f330f5c8acbb80e4dc4e5ba69c9fd5b5855ea804

Image ID:
sha256:955e088a85b5378b00275842bc839eea8cb04ca0782ed79eaa3a967d11fd22e5

vLLM package:
0.26.1rc0+infernal.invocation.cu133.r18.vllmf0fa1ce.b12x75787c7

vLLM source commit:
6dc2f516688fe6f84c6994dcd20fddf296853a6c
vLLM integration tree:
f0fa1cefc1865d316c2478525f550e7646addc40

B12X source commit:
c25cdba2c1df7a69b2d7771e4243e12a8fbf19d5
B12X integration tree:
75787c7a7431b3bea414d2ebf5f2b8671b23eb33

FlashInfer:
0.6.18+cu133, commit 1ac6942776b383c6b03c7a5805a22e72a3e3349f

PyTorch:
2.13.0, CUDA 13.3

LMCache:
0.5.2+glm52dcp.5, integration tree e045d729bc5c4c63a40e13d032f42923de97812f

InstantTensor:
0.1.9, commit 49b4010afc1cae0441e71fe0b0bffc24fa05e932
```

### Model

```text
Model: DeepSeek-V4-Flash-0731
Architecture: DeepseekV4ForCausalLM
model_type: deepseek_v4
Native dspark_block_size: 5
Checkpoint source: local read-only copy of the official checkpoint
```

Selected local checkpoint file hashes:

```text
config.json
6c8f3d2d3b48707541b88f32f22ef3f0f8a6b57d8523281e2b8d3cdb0ae9a023

model.safetensors.index.json
98efab455cf08dfbbbaaba6f570e1bf10bf927d2b4c3c453a59c2f6f0e3be92b

tokenizer_config.json
6ac8c8dc065ed118161d02dd532749ae3f52c243deac27872134fae2f50d8547
```

## Effective launch command

```bash
vllm serve /root/models/DeepSeek-V4-Flash-0731 \
  --served-model-name DeepSeek-V4-Flash-0731 \
  --host 0.0.0.0 \
  --port 5001 \
  --trust-remote-code \
  --kv-cache-dtype fp8 \
  --block-size 256 \
  --load-format instanttensor \
  --tensor-parallel-size 4 \
  --decode-context-parallel-size 1 \
  --gpu-memory-utilization 0.972 \
  --max-model-len 1048576 \
  --max-num-seqs 8 \
  --max-num-batched-tokens 8192 \
  --max-cudagraph-capture-size 48 \
  --compilation-config '{"cudagraph_mode":"FULL_AND_PIECEWISE","custom_ops":["all"]}' \
  --async-scheduling \
  --no-scheduler-reserve-full-isl \
  --enable-chunked-prefill \
  --tokenizer-mode deepseek_v4 \
  --tool-call-parser deepseek_v4 \
  --reasoning-parser deepseek_v4 \
  --enable-auto-tool-choice \
  --enable-prompt-tokens-details \
  --enable-force-include-usage \
  --enable-request-id-headers \
  --default-chat-template-kwargs.thinking=true \
  --default-chat-template-kwargs.reasoning_effort=high \
  --enable-flashinfer-autotune \
  --enable-prefix-caching \
  --speculative-config '{"model":"/root/models/DeepSeek-V4-Flash-0731","method":"dspark","num_speculative_tokens":5,"draft_sample_method":"probabilistic","rejection_sample_method":"standard"}' \
  --attention-backend B12X_MLA_SPARSE \
  --moe-backend b12x \
  --linear-backend b12x \
  --kv-transfer-config '{"kv_connector":"LMCacheMPConnector","kv_role":"kv_both","kv_connector_extra_config":{"lmcache.mp.host":"tcp://127.0.0.1","lmcache.mp.port":5555,"lmcache.mp.mq_timeout":60,"lmcache.mp.heartbeat_interval":5,"lmcache.mp.mp_transfer_mode":"auto"}}'
```

Relevant wrapper settings:

```text
MODE=dspark
BACKEND=b12x-a8
TP_SIZE=4
DCP_SIZE=1
DSPARK_TOKENS=5
DRAFT_SAMPLE_METHOD=probabilistic
GRAPH=auto
MAX_MODEL_LEN=1048576
MAX_NUM_SEQS=8
MAX_NUM_BATCHED_TOKENS=8192
GPU_MEMORY_UTILIZATION=0.972
LMCACHE_MODE=ram
LMCACHE_L1_GB=48
LOAD_FORMAT=instanttensor
INSTANTTENSOR_BACKEND=BUFFERED
```

## Original production incident

A captured Bifrost request bundle contains:

- 472,731 prompt tokens when reconstructed against this runtime
- 48 OpenAI tool definitions
- a long agent/session history
- a corrupt assistant response containing newly generated Latin/CJK token soup

The corrupt response included fragments such as:

```text
corredefined. Ifaidea,算 whateverReady,面/,Ridiculous!v0...
...
The 每秒。");, gone; my amazon s3 twosuite responsereplied...
```

The raw response content is not being posted because the source history contains private operational data. A redacted reproduction payload and synthetic/public-safe evidence can be provided privately.

Source bundle hashes:

```text
bifrost-corrupt-request-bundle.json
9f3d0e0900396491ec958fc00d043a455ad5d5ea6b47f8148b02dfd64e559e5c

captured output_message canonical SHA-256
be64653c99634b9d917e0c1d8693633558378fc41b9e2e34aeff263cb1486991
```

### Request shape

The captured history was recovered into an OpenAI-compatible message array. Because the embedded historical JSON contained invalid backslash escapes, this is an **exact semantic replay**, not a byte-identical transport replay.

Each controlled run used:

```text
Prompt tokens: 472,816
Messages: system + complete retained historical session + one natural long-output request
Tools: 48 real OpenAI tool definitions
Temperature: 1.0
max_tokens: 65,536
Concurrency: 6
Requests: 36
Normal EOS behavior; ignore_eos was not used
Raw SSE captured per request
```

The appended user instruction requested a naturally varied 12,000-word engineering dossier grounded in the retained history. It explicitly prohibited tool calls so each request generated sustained natural language rather than stopping at a tool boundary.

### Reproduction command

The retained harness is currently an internal forensic script; equivalent request behavior is:

```bash
python replay_hermes_session_ds4.py \
  --messages-json messages.json \
  --tools-json tools.json \
  --base-url http://127.0.0.1:5001/v1 \
  --model DeepSeek-V4-Flash-0731 \
  --append-user-text 'Produce a comprehensive engineering dossier of at least 12,000 words using the complete session history. Reconstruct chronology, architecture, constraints, every decision, competing root-cause hypothesis, all evidence, failure modes, operational procedures, rollback paths, validation matrix, performance analysis, risk register, and detailed phased implementation plan. Use varied original prose and substantial sections. Continue naturally until the dossier is complete. Do not call tools.' \
  --max-tokens 65536 \
  --temperature 1.0 \
  --concurrency 6 \
  --repeats 36 \
  --output-dir evidence
```

The harness sends standard streaming `/v1/chat/completions` requests and saves every raw SSE byte plus parsed `delta.reasoning`, `delta.reasoning_content`, `content`, tool deltas, finish reason, usage, and timings.

## Result: reproduction 1

```text
Requests: 36
Completion tokens: 634,181
Median completion: 18,131 tokens
Range: 2,235–21,901 tokens
Longest request: 296.1 seconds
Transport errors: 0
```

New foreign/glued token not present in the input history:

```text
the toolログ said DONE
```

This occurred in request 14 reasoning after a long coherent English prefix. The replay also produced malformed/unclosed markup findings across long outputs.

DSpark counters over the run window:

```text
Draft tokens: 982,430
Accepted tokens: 453,728
Aggregate acceptance: 46.18%
```

There was no global acceptance collapse.

Evidence hashes:

```text
request-0014/raw.sse
4fbe94e498e45e4451471c0defc58df4a5bb9a6ed42251af8290477ff17cff76

request-0014/parsed.json
bf59e9521624b633cbbf807b095d290e6097e000647bd61805e6885909e84e2d
```

## Result: reproduction 2

A second independent invocation used the same payload and runtime:

```text
Requests: 36
Completion tokens: 709,040
Median completion: 19,557 tokens
Range: 14,715–25,850 tokens
Longest request: 257.2 seconds
Transport errors: 0
```

New foreign/glued strings absent from the input history appeared in multiple responses:

```text
request 17: "キャラ"
request 28: "w聾ired it to our own GPU box"
request 31: "a documented p большой pitfall"
```

Other CJK excerpts that quoted corruption already present in the retained history were excluded from new-output evidence.

DSpark counters over the run window:

```text
Draft tokens: 1,106,740
Accepted tokens: 502,841
Aggregate acceptance: 45.43%
```

Evidence hashes:

```text
request-0017/raw.sse
1e87ea85d8d161285a7bcdae4cb8b5df887d06befef0cd3772ec811f9f5625c8

request-0028/raw.sse
06d628bca0c8824ff46bb7fb1da3d81813c0d89c625ff1d89eb513547958da45

request-0031/raw.sse
146e046b77a5e42d23ff85531b0514d343d75f19c6e0e4ba9fbc7129eb3c9bdc
```

### Exact captured request replay, ordinary output lengths

- Serial: 3/3 clean
- Concurrency 3: 9/9 clean
- Concurrency 6: 30/30 clean

These requests generally ended after hundreds to a few thousand output tokens. Prompt identity alone is insufficient.

### Target-only, no speculation

The same environment was recreated with `MODE=dspark-mtp0`, no speculative config, while retaining TP4, 1M context, prefix caching, and LMCache.

Test matrix included:

- real 186k-token session: 3/3 clean
- real 324k-token session: 30/30 clean, including concurrency 8
- 350k prompt, concurrency 3: 6/6 clean
- 500k prompt, concurrency 4: 8/8 clean
- 500k prompt, concurrency 6: 12/12 clean
- 816,705-token varied prompt, concurrency 4: 12/12 clean
- 816,705-token varied prompt, concurrency 6: 60/60 clean
- 4 rounds of six live threads plus six disjoint pressure histories, local prefix-cache reset, and LMCache external replay: 72 requests / 28.65M prompt tokens / 136,626 natural completion tokens

That target-only churn reproduced CUDA allocation failures and one isolated foreign character, but not repeatable new cross-script corruption. DSpark made the cross-script phenotype repeatable under the retained long-output workload.

## What this establishes

1. The output corruption is generated server-side; it exists in raw SSE.
2. It is not explained by HTTP transport, client rendering, or tokenizer decoding alone.
3. A long contaminated input history is not sufficient by itself; short exact replays are clean.
4. Long natural generation with DSpark and concurrency 6 reproduces new foreign/glued tokens in two independent 36-request invocations.
5. Aggregate draft acceptance does not need to collapse globally: both reproductions occurred at approximately 45–46% acceptance.
6. The issue is intermittent at the individual-request level and is more likely after thousands of decode steps.
7. The target-only control is dramatically harder to trigger and did not reproduce the same repeated cross-script class in the retained matrix.

This differs from rtx6kpro#53/#72 in that corruption can appear in isolated responses without a batch-wide acceptance collapse. It is closer to local-inference-lab/vllm#475 in being isolated and stochastic, but the phenotype here is new cross-script token insertion rather than repeated four-grams.

## Additional observations

- During some high-context runs, the CUDA caching allocator logged failed transient allocations while Docker remained `OOMKilled=false` and health continued to pass. These warnings did not occur in the two retained long-output reproduction windows, so they are not required for this reproduced phenotype.
- The retained input history already contains quoted examples of earlier corruption. For this report, only exact foreign/glued strings absent from the complete input were counted as new corruption.
- Tool and reasoning parser output was preserved separately from visible content to rule out client-side mixing.

## Requested guidance

1. Which DSpark/B12X state should be instrumented per request to identify the first divergent draft or target token?
2. Can per-request accepted/drafted tokens and verifier logits be recorded without materially changing scheduling?
3. Are there known r18 issues in DSpark probabilistic sampling, ragged multi-request verification, B12X sparse MLA decode, or CUDA graph replay that can emit isolated foreign tokens without aggregate acceptance collapse?
4. Which current integration tree or patch should be tested first against this retained workload?
5. Would maintainers prefer a sanitized synthetic bundle, the redacted authentic semantic payload, full raw SSE files, or an instrumented rerun?

## Security / privacy note

The full authentic history contains private operational content and is intentionally not attached publicly. The exact hashes, runtime contract, reproduction results, and selected minimal corrupt excerpts are included above. We can provide a sanitized synthetic reproducer and private evidence bundle to maintainers on request.

### Before submitting

- [x] Searched this repository for DeepSeek-V4, DSpark, multilingual, garbled, and token-soup issues.
- [x] Compared against local-inference-lab/vllm#475 and local-inference-lab/rtx6kpro#53/#72.
- [x] Compared against upstream vllm-project/vllm#50720 and #52938.

**Comment by devinkuhn** [comment]
Additional production evidence after filing:

Two new real Slack/agent responses showed the full catastrophic phenotype again while this exact r18 DSpark K5 runtime remained active.

Observed examples:

```text
QraofA| ... AISLICL</ | DSML | parameter>
< | DSML | parameter name="path" string="true">/tmp/junk.patch</ | DSML | parameter>
</ | DSML | invoke>
```

and a separate response that began coherently, then changed to:

```text
coincidence has been getting a of:000,
which is, that is a emptied behind out of / pro
...</blockquote,
...</para still_bilmkaga bir choose. "artir. We
dwells the validation of anlama vernis...
```

The client/model then self-commented that the output was nonsense/glitched and attempted to resume. This is the original sustained word-salad/DSML-fragment phenotype, not merely the isolated foreign-token precursor from the controlled harness.

Runtime correlation:

- `MODE=dspark`, K5 probabilistic, GMU 0.972
- cumulative DSpark acceptance at inspection: 47.87%
- transient CUDA allocation failures in the same production window:
  - 88 MiB request with ~85–87 MiB free on ranks 1/2
  - 96 MiB request with ~27 MiB free on rank 0 and ~65–67 MiB free on ranks 1/2
- Docker remained `OOMKilled=false`, zero restarts, and requests completed normally
- Prefix-cache hit rate around 98.6–98.7%; external-prefix hit rate around 4%

Screenshots are retained privately because they include operational Slack context. Sanitized crops/raw request IDs can be supplied to maintainers.
